### ISSUE TITLE
Cannot install because failed at the hashring@3.1.1 install script 'node-gyp rebuild'.

### ISSUE THREAD

**Issue by zsxsoft** [3rd-Eden/memcached#265]
```
System: Ubuntu 14.04 x64
Command: npm install memcached --verbose
npm info using npm@2.13.3
npm info using node@v3.2.0
```

I tried to install ``memcached@2.1.0``, but failed. Then I tested to install ``hashring@3.1.1`` manually and predictably failed. Afterward, I forced to upgrade the version of ``hashring`` to 3.2.0, then ``memcached`` installed successfully.

After that, I create a ``Dockerfile`` based on [iojs:3.2.0](https://github.com/nodejs/docker-iojs/blob/master/3.2/Dockerfile) to install it. It failed, too.

So I don't know what should I do. Is it necessary for me to do something to solve this problem? Or should I wait for you to update? Thanks.

Log here:
```
npm info linkStuff hashring@3.1.1
npm verb linkBins hashring@3.1.1
npm verb linkMans hashring@3.1.1
npm verb rebuildBundles hashring@3.1.1
npm verb rebuildBundles [ 'bindings', 'connection-parse', 'nan', 'simple-lru-cache' ]
npm info install hashring@3.1.1

> hashring@3.1.1 install /home/zsx/danmu-server/node_modules/memcached/node_modules/hashring
> node-gyp rebuild

gyp info it worked if it ends with ok
gyp verb cli [ '/bin/iojs',
gyp verb cli   '/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js',
gyp verb cli   'rebuild' ]
gyp info using node-gyp@2.0.2
gyp info using node@3.2.0 | linux | x64
gyp verb command rebuild []
gyp verb command clean []
gyp verb clean removing "build" directory
gyp verb command configure []
gyp verb check python checking for Python executable "python2" in the PATH
gyp verb `which` succeeded python2 /usr/bin/python2
gyp verb check python version `python2 -c "import platform; print(platform.python_version());"` returned: "2.7.6\n"
gyp verb get node dir no --target version specified, falling back to host node version: v3.2.0
gyp verb command install [ 'v3.2.0' ]
gyp verb install input version string "v3.2.0"
gyp verb install installing version: 3.2.0
gyp verb install --ensure was passed, so won't reinstall if already installed
gyp verb install version is already installed, need to check "installVersion"
gyp verb got "installVersion" 9
gyp verb needs "installVersion" 9
gyp verb install version is good
gyp verb get node dir target node version installed: 3.2.0
gyp verb build dir attempting to create "build" dir: /home/zsx/danmu-server/node_modules/memcached/node_modules/hashring/build
gyp verb build dir "build" dir needed to be created? /home/zsx/danmu-server/node_modules/memcached/node_modules/hashring/build
gyp verb build/config.gypi creating config file
gyp verb build/config.gypi writing out config file: /home/zsx/danmu-server/node_modules/memcached/node_modules/hashring/build/config.gypi
gyp verb config.gypi checking for gypi file: /home/zsx/danmu-server/node_modules/memcached/node_modules/hashring/config.gypi
gyp verb common.gypi checking for gypi file: /home/zsx/danmu-server/node_modules/memcached/node_modules/hashring/common.gypi
gyp verb gyp gyp format was not specified; forcing "make"
gyp info spawn python2
gyp info spawn args [ '/lib/node_modules/npm/node_modules/node-gyp/gyp/gyp_main.py',
gyp info spawn args   'binding.gyp',
gyp info spawn args   '-f',
gyp info spawn args   'make',
gyp info spawn args   '-I',
gyp info spawn args   '/home/zsx/danmu-server/node_modules/memcached/node_modules/hashring/build/config.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/lib/node_modules/npm/node_modules/node-gyp/addon.gypi',
gyp info spawn args   '-I',
gyp info spawn args   '/home/zsx/.node-gyp/3.2.0/include/node/common.gypi',
gyp info spawn args   '-Dlibrary=shared_library',
gyp info spawn args   '-Dvisibility=default',
gyp info spawn args   '-Dnode_root_dir=/home/zsx/.node-gyp/3.2.0',
gyp info spawn args   '-Dnode_gyp_dir=/lib/node_modules/npm/node_modules/node-gyp',
gyp info spawn args   '-Dmodule_root_dir=/home/zsx/danmu-server/node_modules/memcached/node_modules/hashring',
gyp info spawn args   '--depth=.',
gyp info spawn args   '--no-parallel',
gyp info spawn args   '--generator-output',
gyp info spawn args   'build',
gyp info spawn args   '-Goutput_dir=.' ]
gyp verb command build []
gyp verb build type Release
gyp verb architecture x64
gyp verb node dev dir /home/zsx/.node-gyp/3.2.0
gyp verb `which` succeeded for `make` /usr/bin/make
gyp info spawn make
gyp info spawn args [ 'V=1', 'BUILDTYPE=Release', '-C', 'build' ]
make: Entering directory `/home/zsx/danmu-server/node_modules/memcached/node_modules/hashring/build'
  g++ '-DNODE_GYP_MODULE_NAME=hashvalue' '-D_LARGEFILE_SOURCE' '-D_FILE_OFFSET_BITS=64' '-DBUILDING_NODE_EXTENSION' -I/home/zsx/.node-gyp/3.2.0/include/node -I/home/zsx/.node-gyp/3.2.0/src -I/home/zsx/.node-gyp/3.2.0/deps/uv/include -I/home/zsx/.node-gyp/3.2.0/deps/v8/include -I../node_modules/nan  -fPIC -pthread -Wall -Wextra -Wno-unused-parameter -m64 -O3 -ffunction-sections -fdata-sections -fno-omit-frame-pointer -fno-rtti -fno-exceptions -std=gnu++0x -MMD -MF ./Release/.deps/Release/obj.target/hashvalue/src/hashvalue.o.d.raw  -c -o Release/obj.target/hashvalue/src/hashvalue.o ../src/hashvalue.cc
In file included from ../src/hashvalue.cc:1:0:
../node_modules/nan/nan.h:261:25: error: redefinition of ‘template<class T> v8::Local<T> _NanEnsureLocal(v8::Local<T>)’
 NAN_INLINE v8::Local<T> _NanEnsureLocal(v8::Local<T> val) {
                         ^
../node_modules/nan/nan.h:256:25: error: ‘template<class T> v8::Local<T> _NanEnsureLocal(v8::Handle<T>)’ previously declared here
 NAN_INLINE v8::Local<T> _NanEnsureLocal(v8::Handle<T> val) {
                         ^
../node_modules/nan/nan.h:661:13: error: ‘node::smalloc’ has not been declared
     , node::smalloc::FreeCallback callback
             ^
../node_modules/nan/nan.h:661:35: error: expected ‘,’ or ‘...’ before ‘callback’
     , node::smalloc::FreeCallback callback
                                   ^
../node_modules/nan/nan.h: In function ‘v8::Local<v8::Object> NanNewBufferHandle(char*, size_t, int)’:
../node_modules/nan/nan.h:665:50: error: ‘callback’ was not declared in this scope
         v8::Isolate::GetCurrent(), data, length, callback, hint);
                                                  ^
../node_modules/nan/nan.h:665:60: error: ‘hint’ was not declared in this scope
         v8::Isolate::GetCurrent(), data, length, callback, hint);
                                                            ^
../node_modules/nan/nan.h: In function ‘v8::Local<v8::Object> NanNewBufferHandle(const char*, uint32_t)’:
../node_modules/nan/nan.h:672:67: error: call of overloaded ‘New(v8::Isolate*, const char*&, uint32_t&)’ is ambiguous
     return node::Buffer::New(v8::Isolate::GetCurrent(), data, size);
                                                                   ^
../node_modules/nan/nan.h:672:67: note: candidates are:
In file included from ../node_modules/nan/nan.h:25:0,
                 from ../src/hashvalue.cc:1:
/home/zsx/.node-gyp/3.2.0/include/node/node_buffer.h:31:40: note: v8::MaybeLocal<v8::Object> node::Buffer::New(v8::Isolate*, v8::Handle<v8::String>, node::encoding) <near match>
 NODE_EXTERN v8::MaybeLocal<v8::Object> New(v8::Isolate* isolate,
                                        ^
/home/zsx/.node-gyp/3.2.0/include/node/node_buffer.h:31:40: note:   no known conversion for argument 3 from ‘uint32_t {aka unsigned int}’ to ‘node::encoding’
/home/zsx/.node-gyp/3.2.0/include/node/node_buffer.h:43:40: note: v8::MaybeLocal<v8::Object> node::Buffer::New(v8::Isolate*, char*, size_t) <near match>
 NODE_EXTERN v8::MaybeLocal<v8::Object> New(v8::Isolate* isolate,
                                        ^
/home/zsx/.node-gyp/3.2.0/include/node/node_buffer.h:43:40: note:   no known conversion for argument 2 from ‘const char*’ to ‘char*’
In file included from ../src/hashvalue.cc:1:0:
../node_modules/nan/nan.h: In function ‘v8::Local<v8::Object> NanNewBufferHandle(uint32_t)’:
../node_modules/nan/nan.h:676:61: error: could not convert ‘node::Buffer::New(v8::Isolate::GetCurrent(), ((size_t)size))’ from ‘v8::MaybeLocal<v8::Object>’ to ‘v8::Local<v8::Object>’
     return node::Buffer::New(v8::Isolate::GetCurrent(), size);
                                                             ^
../node_modules/nan/nan.h: In function ‘v8::Local<v8::Object> NanBufferUse(char*, uint32_t)’:
../node_modules/nan/nan.h:683:12: error: ‘Use’ is not a member of ‘node::Buffer’
     return node::Buffer::Use(v8::Isolate::GetCurrent(), data, size);
            ^
make: *** [Release/obj.target/hashvalue/src/hashvalue.o] Error 1
make: Leaving directory `/home/zsx/danmu-server/node_modules/memcached/node_modules/hashring/build'
gyp ERR! build error 
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:269:23)
gyp ERR! stack     at emitTwo (events.js:87:13)
gyp ERR! stack     at ChildProcess.emit (events.js:172:7)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:200:12)
gyp ERR! System Linux 3.16.0-30-generic
gyp ERR! command "/bin/iojs" "/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /home/zsx/danmu-server/node_modules/memcached/node_modules/hashring
gyp ERR! node -v v3.2.0
gyp ERR! node-gyp -v v2.0.2
gyp ERR! not ok 
npm verb unsafe-perm in lifecycle true
npm info hashring@3.1.1 Failed to exec install script
npm verb unlock done using /home/zsx/.npm/_locks/hashring-3d0b7f9d64d172c7.lock for /home/zsx/danmu-server/node_modules/memcached/node_modules/hashring
npm verb about to build /home/zsx/danmu-server/node_modules/memcached
npm verb unlock done using /home/zsx/.npm/_locks/memcached-08236ba7045d70f6.lock for /home/zsx/danmu-server/node_modules/memcached
npm verb stack Error: hashring@3.1.1 install: `node-gyp rebuild`
npm verb stack Exit status 1
npm verb stack     at EventEmitter.<anonymous> (/lib/node_modules/npm/lib/utils/lifecycle.js:214:16)
npm verb stack     at emitTwo (events.js:87:13)
npm verb stack     at EventEmitter.emit (events.js:172:7)
npm verb stack     at ChildProcess.<anonymous> (/lib/node_modules/npm/lib/utils/spawn.js:24:14)
npm verb stack     at emitTwo (events.js:87:13)
npm verb stack     at ChildProcess.emit (events.js:172:7)
npm verb stack     at maybeClose (internal/child_process.js:764:16)
npm verb stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:211:5)
npm verb pkgid hashring@3.1.1
npm verb cwd /home/zsx/danmu-server
npm ERR! Linux 3.16.0-30-generic
npm ERR! argv "/bin/iojs" "/bin/npm" "install" "memcached" "--verbose"
npm ERR! node v3.2.0
npm ERR! npm  v2.13.3
npm ERR! code ELIFECYCLE

npm ERR! hashring@3.1.1 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR! 
npm ERR! Failed at the hashring@3.1.1 install script 'node-gyp rebuild'.
npm ERR! This is most likely a problem with the hashring package,
npm ERR! not with npm itself.
npm ERR! Tell the author that this fails on your system:
npm ERR!     node-gyp rebuild
npm ERR! You can get their info via:
npm ERR!     npm owner ls hashring
npm ERR! There is likely additional logging output above.
npm verb exit [ 1, true ]
npm verb unbuild node_modules/memcached/node_modules/hashring
npm info preuninstall hashring@3.1.1
npm info uninstall hashring@3.1.1
npm verb unbuild rmStuff hashring@3.1.1 from /home/zsx/danmu-server/node_modules
npm verb unbuild rmStuff in /home/zsx/danmu-server/node_modules/memcached/node_modules
npm info postuninstall hashring@3.1.1
npm verb gentlyRm don't care about contents; nuking /home/zsx/danmu-server/node_modules/memcached/node_modules/hashring
npm verb unbuild node_modules/memcached
npm info preuninstall memcached@2.1.0
npm info uninstall memcached@2.1.0
npm verb unbuild rmStuff memcached@2.1.0 from /home/zsx/danmu-server/node_modules
npm info postuninstall memcached@2.1.0
npm verb gentlyRm don't care about contents; nuking /home/zsx/danmu-server/node_modules/memcached

npm ERR! Please include the following file with any support request:
npm ERR!     /home/zsx/danmu-server/npm-debug.log
```

**Comment by avin45h** [comment]
Do you have basic build tools installed? Looks to me that it is failing during make

**Comment by zsxsoft** [comment]
@avin45h 
node-gyp requires:
   - python (2.7.6)
   - make (OK)
   - gcc (4.8.4)
So I think my environment is OK. To the log, it's just compilation errors. Maybe not compatible with iojs@3.2.0.

**Comment by jkrems** [comment]
I think `memcached` just needs a version bump so the latest version on npm picks up `hashring@3.2.0` which is node 3.x+ compatible (it already refers to that version on master, it's just not released).

**Comment by zsxsoft** [comment]
@jkrems I'm agree with you, and the ``harhring@3.2.0`` has been published to npm,

**Comment by 3rd-Eden** [comment]
i'll look in to release a new version today

**Comment by diversario** [comment]
Perfect. This is a blocker for me right now.

**Comment by Qard** [comment]
Same here! :shipit:

**Comment by kitt1987** [comment]
+1

**Comment by 3rd-Eden** [comment]
Published fix as 2.2.0
